splunk-otel-android/src/main/java/com/splunk/rum/SessionId.java

```diff
@@ -60,20 +64,39 @@ String getSessionId() {
                     sessionIdChangeListener.onChange(currentValue, newId);
                 }
             }
-            return value.get();
+            currentValue = value.get();
         }
+        resetTimeout();
         return currentValue;
     }
 
-    void setSessionIdChangeListener(SessionIdChangeListener sessionIdChangeListener) {
-        this.sessionIdChangeListener = sessionIdChangeListener;
+    void startInactivityTimeout() {
+        timeoutStartNanos = clock.nanoTime();
     }
 
     private boolean sessionExpired() {
+        // TODO: it probably should use nanoTime(); now() javadoc explicitly states that it's not meant to be used to compute duration
         long elapsedTime = clock.now() - createTimeNanos;
         return elapsedTime >= SESSION_LIFETIME_NANOS;
     }
 
+    private boolean sessionTimedOut() {
+        if (timeoutStartNanos == NO_TIMEOUT) {
+            // we haven't started the inactivity timeout yet
+            return false;
+        }
+        long elapsedTime = clock.nanoTime() - timeoutStartNanos;
+        return elapsedTime >= SESSION_TIMEOUT_NANOS;
+    }
+
+    private void resetTimeout() {
+        timeoutStartNanos = NO_TIMEOUT;
```

Comment:
it seems like it would be simplest just to keep track of a "last accessed time" that updates whenever anyone asks for a sessionId, and just check for how old it is, rather than having this reset logic in addition.

Reply:
It is possible for an app to be in the foreground and not send any spans for >15 minutes, isn't it? In that case I'd still need another variable (a boolean) that'd be set whenever the app gets backgrounded.

Reply:
I'm not sure how you're handling that now, though? Seems like you're only checking for timeout when the sessionId is accessed already. My idea would just remove one extra bit to check.

Reply:
Hm, but the session is not supposed to time out when the app is active, in the foreground - in that case it's supposed to last full 4 hours (even if the app doesn't send any span for 15 minutes). The timeout should only be applied when the app gets backgrounded.

Reply:
I don't know how we can possibly track that an app is "active" if no activity is being observed by the instrumentation. The best we can do (as far as I know) is to just track "liveness" by the generation of spans. If no traceable activity happens for 15 minutes, could we not consider that as a new session?

Reply:
If an app developer wants to have instrumentation-idle apps to have extended sessions, we could recommend that they generate some sort of "ping" event every few minutes to keep it alive, I guess?

Reply:
@jkwatson PTAL at this PR now - I've talked to Seerut about the background/foreground timeout requirements, and reimplemented the whole thing.

Reply:
ah, cool. As long as our foreground/background detection is reliable, then, 👍🏽